[Game] Coffers

ZeromusXYZ opened the pull request:
Adds coffer/chest functionality to housing

ZeromusXYZ committed:
[Game] Added Coffers

Rearranged the loading order of Doodad Templates to be able to access their funcs before the template is created.

Added DoodadCoffer and DoodadCofferTemplate and changed DoodadManager.Create to create a DoodadCoffer where appropriate.

Currently only assigns Capacity to the Coffer.
ItemContainer is not yet linked.
[Game] WIP - Coffers

Added item_container_id field to doodads MySQL table to be used by coffers.
Added ability to view coffer contents.
Coffer is now locked when somebody is viewing it.
[Game] WIP - CSSwapCofferItems

Added GM command to handle some basic coffer stuff.
Changed internal ContainerId to ulong.
Added null-check to CanAccept
Added/renamed SplitOrMoveItemEx()
Newly created ItemContainers of previously existing coffers should now be properly saved.
Added /coffer GM command to accesslevel.json at Moderator level (50)
[Game] SplitCofferItem

Implemented CSSplitCofferItem.
Fixed a bug with SplitOrMoveItemEx() where when merging items it could potentially send the wrong packet to the client, resulting in a phantom duplicate item on the client side.
[Game] Coffer - RecoverItem

Added InitializeCoffer() to simplify the code
Added a Hack to ignore DoodadFuncCoffer so you can recover it from your house.
Added function to clean up the ItemContainer linked to a coffer when it's removed
Added a virtual AllowRemoval() to check if a object can be picked up by a player (RecoverItem)
[Game] WIP - Coffer Permissions

Added better use permission check functions.
Added cached Account Ids in NameManager.
Partially Implemented SCChangeDoodadData.
Added Data field to doodads MySql.
Added virtual AllowedToInteract() to GameObject
[Game] Added SCDoodadChangedPacket

Implemented SCDoodadChangedPacket which is used to handle changing of coffer permission settings.
Setting coffer permissions now works correctly.

ZeromusXYZ committed:
[Game] Added Coffers

Rearranged the loading order of Doodad Templates to be able to access their funcs before the template is created.

Added DoodadCoffer and DoodadCofferTemplate and changed DoodadManager.Create to create a DoodadCoffer where appropriate.

Currently only assigns Capacity to the Coffer.
ItemContainer is not yet linked.
[Game] WIP - Coffers

Added item_container_id field to doodads MySQL table to be used by coffers.
Added ability to view coffer contents.
Coffer is now locked when somebody is viewing it.
[Game] WIP - CSSwapCofferItems

Added GM command to handle some basic coffer stuff.
Changed internal ContainerId to ulong.
Added null-check to CanAccept
Added/renamed SplitOrMoveItemEx()
Newly created ItemContainers of previously existing coffers should now be properly saved.
Added /coffer GM command to accesslevel.json at Moderator level (50)
[Game] SplitCofferItem

Implemented CSSplitCofferItem.
Fixed a bug with SplitOrMoveItemEx() where when merging items it could potentially send the wrong packet to the client, resulting in a phantom duplicate item on the client side.
[Game] Coffer - RecoverItem

Added InitializeCoffer() to simplify the code
Added a Hack to ignore DoodadFuncCoffer so you can recover it from your house.
Added function to clean up the ItemContainer linked to a coffer when it's removed
Added a virtual AllowRemoval() to check if a object can be picked up by a player (RecoverItem)
[Game] WIP - Coffer Permissions

Added better use permission check functions.
Added cached Account Ids in NameManager.
Partially Implemented SCChangeDoodadData.
Added Data field to doodads MySql.
Added virtual AllowedToInteract() to GameObject
[Game] Added SCDoodadChangedPacket

Implemented SCDoodadChangedPacket which is used to handle changing of coffer permission settings.
Setting coffer permissions now works correctly.

ZeromusXYZ committed:
Merge remote-tracking branch 'AAEmu/develop' into feature/coffer

ZeromusXYZ committed:
[Game] Added Coffers - CodeFactor cleaning

ZeromusXYZ committed:
[Game] Added Coffers - CodeFactor cleaning

RogerBarreto commented on AAEmu.Game/Core/Managers/World/SpawnManager.cs:
```diff
@@ -286,6 +291,7 @@ public void Load()
                             doodad.ItemTemplateId = sourceItem?.TemplateId ?? itemTemplateId;
                             // Grab Ucc from it's old source item
                             doodad.UccId = sourceItem?.UccId ?? 0;
+                            doodad._data = data;
```
Would suggest using the property Data instead of the field

ZeromusXYZ replied:
The reason I don't use the property *here* is because it would trigger a .Save() while it's still loading the doodad. I should probably just make a AssignData() function or so instead.

RogerBarreto commented on AAEmu.Game/Models/Game/DoodadObj/Doodad.cs:
```diff
@@ -54,8 +69,12 @@ public class Doodad : BaseUnit
+        /// <summary>
+        /// Try to use Data instead
+        /// </summary>
+        public int _data;
```
Would make it private. OCP.

ZeromusXYZ replied:
Cannot due to above issue

AAGene commented on AAEmu.Game/Core/Managers/ExpeditionManager.cs:
```diff
@@ -546,5 +546,10 @@ public void SendExpeditions(Character character)
             character.SendPacket(new SCExpeditionRolePolicyListPacket(new List<ExpeditionRolePolicy>()));
         }
+
+        public uint GetExpeditionOfCharacter(uint ownerId)
```
ownerId => characterId

ZeromusXYZ replied:
I guess I'll change that, the naming here was auto-generated, I write the function where I needed it first, and just told Rider to implement it for me, so it named it ownerId because of how it's used.

AAGene commented on AAEmu.Game/Core/Managers/FamilyManager.cs:
```diff
@@ -277,5 +278,10 @@ private static FamilyMember GetMemberForCharacter(Character character, byte owne
             return member;
         }
+
+        public uint GetFamilyOfCharacter(uint ownerId)
```
ownerId => characterId

AAGene commented on AAEmu.Game/Core/Packets/G2C/SCDoodadOriginatorPacket.cs:
```diff
@@ -7,18 +7,20 @@ public class SCDoodadOriginatorPacket : GamePacket
             stream.WriteBc(_objId);
             stream.Write(_newOwnerId);
-            stream.Write((uint)0x00000000); // UCC, Guild, Faction, something else ?
+            stream.Write(_unknown); // UCC, Guild, Faction, something else ?
```
unknown => this is a faction ID

AAGene commented on AAEmu.Game/Models/Game/DoodadObj/Doodad.cs:
```diff
@@ -97,6 +116,10 @@ public void Use(Unit unit, uint skillId, uint recursionDepth = 0)
                         continue;
+                    
+                    // TODO: Hack!!
+                    if (func.FuncType == "DoodadFuncCoffer")
```
You should be able to implement the DoodadFuncCoffer class and make it simply return no next phase

AAGene commented on AAEmu.Game/Models/Game/DoodadObj/Doodad.cs:
```diff
@@ -326,13 +349,18 @@ public void Save()
                         parentDoodadId = pDoodad.DbId;
+                    // Check if coffer
+                    ulong itemContainerId = 0u;
+                    if (this is DoodadCoffer coffer)
```
I would have preferred something along the lines of having DoodadCoffer implement the coffer-specific code in its own Delete method. 

I do see why it's not done that way tho :/

ZeromusXYZ replied:
I'll implement it in another way

AAGene commented on AAEmu.Game/Models/Game/Items/Containers/CofferContainer.cs:
```diff
@@ -44,9 +46,23 @@ private bool CanAcceptTemplate(ItemTemplate itemTemplate)
+                   base.CanAccept(item, targetSlot));
+        }
+
+        public override void Delete()
```
Would this behavior later on be replaced by mailing items to the owner?

ZeromusXYZ replied:
This would need to be part of a Housing update I still need to do yes. But I felt it was not related enough to be put in this coffer PR itself.
I'll address some smaller house issues next, and will add it there. This code is just here to kill any remaining items that might have been left in there.

AAGene reviewed (commented):
Great job overall, some minor changes I've noticed but nothing that prevents merging IMO

ZeromusXYZ committed:
[Game] Coffer - Suggested changes

Renamed ownerId to characterId for GetxxxOfCharacter().
Added SetData() to circumvent directly accessing Doodad._data.
Fixed some typos for warnings.
Renamed _unknown to _faction for SCDoodadOriginatorPacket
Removed CofferDbId getter as it's replaced with GetItemContainerId()

ZeromusXYZ committed:
[Game] Coffer - Suggested changes

Renamed ownerId to characterId for GetxxxOfCharacter().
Added SetData() to circumvent directly accessing Doodad._data.
Fixed some typos for warnings.
Renamed _unknown to _faction for SCDoodadOriginatorPacket
Removed CofferDbId getter as it's replaced with GetItemContainerId()

ZeromusXYZ committed:
[Game] Coffer - Mail back items on house demolish

ZeromusXYZ committed:
[Game] Coffer - Mail back items on house demolish

ZeromusXYZ committed:
[Game] Fix: Coffer RecoverItem, Return items on Demolish

Adjusted Doodad.Use() to be able to restrict it's function to only one funcType.
Implemented DoodadFuncCoffer for completeness sake. Normally this should never actually get called.
Moved were coffer items get restored, for now it will restore **all** items to the original owner of the coffer.

ZeromusXYZ committed:
[Game] Fix: Coffer RecoverItem, Return items on Demolish

Adjusted Doodad.Use() to be able to restrict it's function to only one funcType.
Implemented DoodadFuncCoffer for completeness sake. Normally this should never actually get called.
Moved were coffer items get restored, for now it will restore **all** items to the original owner of the coffer.

ZeromusXYZ committed:
[Game] Fixed RecoverItem

Did a proper fix for RecoverItem triggering DoodadFuncRecoverItem without having to resort to weird limiting hacks.

ZeromusXYZ committed:
[Game] Fixed RecoverItem

Did a proper fix for RecoverItem triggering DoodadFuncRecoverItem without having to resort to weird limiting hacks.

rogerbarreto committed:
features/net6.0 upgrade (#467)

* Converting from .net 3.1 to 6.0 + error handling

* Changed more files to net6 and adjusted game abort messages and behavior

* Validated Docker for net6.0 and some improvements and warning fixes from newer packages versions

* Minor codefactor fix

* Changed dotnet core action to net6

* Removing reference to config.json in the csproj files

* Updated the action to .net6 and fixed missing warnings

* Wrong identation with

* Updated all packages from AAEmu.Tests

* Comment fix and some other small changes

ZeromusXYZ committed:
Merge remote-tracking branch 'origin/feature/coffer' into feature/coffer
